### PR TITLE
Group orders per activity

### PR DESCRIPTION
**Pull request by wilco375**
See #177 

I create two separate components for the order screen of an activity and the order history screen of a user. The history screen of the user groups orders per activity, and only loads the orders when a user clicks on an activity which causes the orders to load much faster when a user has a large number of orders.

### CODE REVIEW

File: app/views/users/show.html.slim [hunk 1]
```diff
@@ -92,7 +92,7 @@
   hr
 
   .col-sm-12
-    .row.card
+    .row.card.d-block
```
Comment (wilco375) [inline app/views/users/show.html.slim]: For some reason there is a large padding at the bottom of the card after expanding an order in the user history overview. Changing the card from a flexbox to a block fixes this somehow. If anybody can find a better fix for or the origin of this issue, be my guest.

File: app/controllers/orders_controller.rb [hunk 2]
```diff
@@ -65,7 +65,7 @@ def proper_json
       include: { order_rows: {
         only: %i[id product_count price_per_product],
         include: { product: { only: %i[id name] } }
-      }, user: { only: :name }, activity: { only: :title } } }
```
Comment (wilco375) [inline app/controllers/orders_controller.rb]: The activity field is no longer needed since it wasn't shown in the activity order view and it is no longer shown in the user order view since the activities are loaded in first via a separate endpoint.

File: .gitignore [hunk 3]
```diff
@@ -41,3 +41,6 @@ yarn-debug.log*
 /yarn-error.log
 yarn-debug.log*
 .yarn-integrity
+
+# Ignore Jetbrains IDE files
+/.idea
```
Comment (wilco375) [inline .gitignore]: Kept accidentally adding my .idea folder from RubyMine... I could make a separate PR for this but I hope this is fine too.

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -68,6 +68,15 @@ def api_token # rubocop:disable Metrics/AbcSize
     @token = JSON.parse(token_response)['access_token']
   end
 
+  def activities
+    user = User.find(params.require(:id))
+    authorize user
+
+    activities = Activity.select(%i[id title start_time]).joins(:orders).merge(Order.orders_for(user)).distinct
```
Comment (wilco375) [inline app/controllers/users_controller.rb]: This query gets all the activities that the given user ordered something at.

File: app/javascript/components/activityorderhistory.vue [hunk 5]
```diff
@@ -0,0 +1,64 @@
+<template lang="html">
```
Comment (wilco375) [inline app/javascript/components/activityorderhistory.vue]: I made this a separate component instead of integrating this in `orderhistory.vue` because the isLoading and orders attributes would have to be arrays otherwise, which isn't very nice.

File: app/javascript/components/activityorders.vue [hunk 6]
```diff
@@ -0,0 +1,126 @@
+<template lang="html">
```
Comment (wilco375) [inline app/javascript/components/activityorders.vue]: This is the order component shown in the activity stripe system

File: app/javascript/components/orderhistory.vue [hunk 7]
```diff
@@ -0,0 +1,92 @@
+<template lang="html">
```
Comment (wilco375) [inline app/javascript/components/orderhistory.vue]: This is the order component shown on the user page

File: app/javascript/orderscreen/orderhistory.vue [hunk 8]
```diff
@@ -1,227 +0,0 @@
-<template lang="html">
```
Comment (wilco375) [inline app/javascript/orderscreen/orderhistory.vue]: This component has been split up into `components/orderhistory.vue` and `components/activityorders.vue`

File: app/policies/user_policy.rb [hunk 9]
```diff
@@ -14,4 +14,8 @@ def search?
   def show?
     user&.treasurer? || record == user
   end
+
+  def activities?
+    user&.treasurer? || record == user
```
Comment (wilco375) [inline app/policies/user_policy.rb]: You can only see the activities you ordered something on of your own, or if you are the treasurer you can see it for everyone.

File: app/views/users/show.html.slim [hunk 10]
```diff
@@ -117,7 +117,7 @@
                     th scope='col'= '#'
                     th scope='col'= 'Datum'
                     th scope='col'= 'Beschrijving'
-                    th scope='col'= 'Bedrag'
+                    th.text-right scope='col'= 'Bedrag'
```
Comment (wilco375) [inline app/views/users/show.html.slim]: The amount column of orders was also aligned to the right. This looks nicer in my opinion so I also aligned the amount column of credit mutations to the right.

File: config/routes.rb [hunk 11]
```diff
@@ -15,6 +15,9 @@
       get :refresh_user_list
       post :search
     end
+    member do
+      get :activities
```
Comment (wilco375) [inline config/routes.rb]: Add the endpoint /users/<id>/activities to get a list of activities that a user ordered something at.

File: README.md [hunk 12]
```diff
@@ -86,3 +86,4 @@ When submitting a PR, it'll need to be approved, but once approved (and green) y
 - [@cmitz](https://github.com/cmitz)
 - [@Matthijsy](https://github.com/Matthijsy)
 - [@cpbscholten](https://github.com/cpbscholten)
+- [@wilco375](https://github.com/wilco375)
```
Comment (wilco375) [inline README.md]: Of course I also want to be in the contributors list :)

File: app/policies/user_policy.rb (hunk 9 above)
Comment (Matthijsy) [inline app/policies/user_policy.rb]: Activities is a plural right? So it show multile records, in that case record==user should not work. I think you should create a scope which limits the records to the current user records

Comment (cmitz) [reply app/policies/user_policy.rb]: We've worked on it together, yesterday. We could not figure out a perspective where using a policy Scope could work.

The reasoning is this:
- You want all activities where queried user (say: Alice) has orders
- The querying user (say: Bob) can only see his own orders, except when Bob is a Treasurer
- A scope would imply that Bob automatically gets his own (👍 ), but a treasurer would receive ALL orders instead of just Alices' (👎 ).
- So: a scope would return too much orders

@wilco375 maybe, just maybe, using the already available scope from Orders (`policy_scope(Order)`) and then further narrow the query down could work

Comment (cmitz) [reply app/policies/user_policy.rb]: I'm thinking:
```ruby
Activity.select(%i[id title start_time]).joins(:orders).merge(policy_scope(Order).where(user: user).distinct
```

I don't know if the syntax is totally correct, though...

File: app/views/users/show.html.slim (hunk 1 above)
Comment (cmitz) [reply app/views/users/show.html.slim]: Als je de table de class `mb-0` geeft, en de `p-2` class van `tab-content` weghaalt, dan heb je hetzelfde effect.

**Comment by cmitz** [comment]
I would like to wait for Wilco's reaction on some of the comments

File: app/policies/user_policy.rb (hunk 9 above)
Comment (cmitz) [reply app/policies/user_policy.rb]: @wilco375 Well I was actually waiting for a comment on this one ;)

Comment (Matthijsy) [reply app/policies/user_policy.rb]: This actually is okay this way. The policy is about the user (since it is a user endpoint) so this is correct, I misunderstood in the first place
